Add transition period for continuity

gershido opened the pull request:
Proposed elections flow:
<img width="901" alt="Screenshot 2024-08-12 at 16 16 20" src="https://github.com/user-attachments/assets/5e641e46-7afa-4e0b-a0ec-9c38f9cffb6b">

gershido committed:
draft

spengrah commented:
@gershido I haven't dug into the code in more detail than a cursory review, but overall this approach looks good to me. The diagram is helpful.

Only thing I'd consider is adding an option to set a new transition period as an arg in `reelection`.

gershido committed:
implementation and tests

gershido committed:
format

spengrah commented on test/JokeraceEligibility.t.sol:
```diff
@@ -20,15 +20,15 @@ contract DeployImplementationTest is DeployImplementation, Test {
 
   uint256 public fork;
-  uint256 public BLOCK_NUMBER = 9_713_194; // the block number where hats module factory was deployed on Goerli;
+  uint256 public BLOCK_NUMBER = 6_488_268; // the block number where hats module factory was deployed on Goerli;
```
sepolia, i assume? If so, please change the comment

spengrah commented on src/JokeraceEligibility.sol:
```diff
@@ -228,7 +242,10 @@ contract JokeraceEligibility is HatsEligibilityModule {
     allowed = block.timestamp >= termEnd
-      || GovernorCountingSimple(payable(underlyingContest)).state() == Governor.ContestState.Canceled;
+      || (
+        currentContest != address(0)
```
When would currentContest == 0? Is it only before the results from the first contest have been pulled?

gershido replied:
Yes

gershido committed:
emit event on election pulling and update logic of allowed reelection time

gershido commented:
Added the following changes:
- `ElectionResultsPulled` event. Currently there's no emitted event when the `pullElectionResults` is successfully called. Adding the event will help indexing the module.
- Added a condition for the `reelectionAllowed` function, for the case in which we're in a transition period, the next contest was set but then for some reason was cancelled.

spengrah commented on src/JokeraceEligibility.sol:
```diff
@@ -31,6 +31,8 @@ contract JokeraceEligibility is HatsEligibilityModule {
   /// @notice Emitted when a reelection is set
   event NewTerm(address NewContest, uint256 newTopK, uint256 newTermEnd, uint256 newTransitionPeriod);
+  /// @notice Emitted when election's results are pulled
+  event ElectionResultsPulled(address NewContest);
```
would like to see a test for this event

spengrah commented on src/JokeraceEligibility.sol:
```diff
@@ -241,10 +244,15 @@ contract JokeraceEligibility is HatsEligibilityModule {
       || (
         currentContest != address(0)
           && GovernorCountingSimple(payable(currentContest)).state() == Governor.ContestState.Canceled
+      ) // or if we're in a transition period, and the next contest was canceled
```
Is there a test for this new logic? I might be missing it, but I don't see one

gershido replied:
Test added

gershido committed:
add tie indication in event and test events

gershido commented:
Added:
- New `isTie` param in the `ElectionResultsPulled` event, and emitting the event for the tie scenario also
- Event tests

spengrah reviewed (approved):
looks good! thanks for adding those tests

gershido committed:
implementation updates

gershido committed:
documentation and optimizations

gershido committed:
documentation

gershido commented:
Following last week's discussion about the election flow, changed the implementation to allow setting up the next term even before the current term has ended. The main implication on the implementation side was that it requires to keep the full set of details for the next term (contest, term end, top K and transition period). Here are the main changes:
- Terms are saved as indexed `TermDetails` structs. There's a `currentTermIndex` variable. The next term will be always saved in `currentTermIndex + 1`.
- Name changes: `reelection` function changed to `setNextTerm`, `pullElectionResults` changed to `startNextTerm`
- Mapping of eligible wearers was changed from `eligibleWearesPerContest` to `eligibleWearersPerTerm`

spengrah commented on src/JokeraceEligibility.sol:
```diff
@@ -29,8 +31,10 @@ contract JokeraceEligibility is HatsEligibilityModule {
-  /// @notice Emitted when a reelection is set
-  event NewTerm(address NewContest, uint256 newTopK, uint256 newTermEnd);
+  /// @notice Emitted when the next term is set
+  event NextTermSet(address NewContest, uint256 newTopK, uint256 newTermEnd, uint256 newTransitionPeriod);
```
style: camel case for `newContest`

spengrah commented on src/JokeraceEligibility.sol:
```diff
@@ -64,18 +68,34 @@ contract JokeraceEligibility is HatsEligibilityModule {
+    /// @notice Term's ending time (a unix timestamp)
+    uint256 termEnd;
+    /// @notice Period of time after the term end when previous elected members are still considered eligible until a
+    /// new term begins.
```
style: join these two comment lines

gershido replied:
It's automatically formatted like that, the `line_length` in the foundry.toml file is set to 120.

spengrah replied:
Oh whoops, I misinterpreted github's ui 🫠

spengrah commented on src/JokeraceEligibility.sol:
```diff
@@ -64,18 +68,34 @@ contract JokeraceEligibility is HatsEligibilityModule {
+    /// @notice Jokerace contest (election)
+    address contest;
+    /// @notice First K winners of the contest will be eligible
+    uint96 topK;
```
nice storage slot packing :)

spengrah commented on src/JokeraceEligibility.sol:
```diff
@@ -135,39 +156,43 @@ contract JokeraceEligibility is HatsEligibilityModule {
-  function pullElectionResults() public returns (bool success) {
-    GovernorCountingSimple currentContest = GovernorCountingSimple(payable(underlyingContest));
+  function startNextTerm() public {
+    TermDetails memory nextTerm = terms[currentTermIndex + 1];
```
Since this function only needs `topK` and `contest` from the `nextTerm` object, it will be cheaper here to set `nextTerm` as a storage pointer rather than SLOADing the entire nextTerm object into memory

ie, 
```solidity
TermDetails stoarge nextTerm = terms[currentTermIndex + 1];
```

gershido replied:
<img width="693" alt="Screenshot 2024-08-27 at 14 42 30" src="https://github.com/user-attachments/assets/ec9f087c-3e1b-409e-8048-2c52c5a7b264">

All the details from the `nextTerm` object are emitted in the event at the end of the function. A different approach would be to just include the next term's index instead of all the details. That's a bit less good for whoever listens to the event because it will require a call to the contract in order to get the details of this new term. What do you think?

spengrah replied:
ah good point, i missed the event params! Keep it as is, then.

spengrah commented on src/JokeraceEligibility.sol:
```diff
@@ -135,39 +156,43 @@ contract JokeraceEligibility is HatsEligibilityModule {
-          address candidate = _getCandidate(currentContest, proposalsOfCurrentRank[proposalIndex]);
-          eligibleWearersPerContest[candidate][address(currentContest)] = true;
+          address candidate = _getCandidate(contest, proposalsOfCurrentRank[proposalIndex]);
+          eligibleWearersPerTerm[candidate][currentTermIndex + 1] = true;
```
this is the second SLOAD of `currentTermIndex`, so I'd suggest storing this in memory

spengrah commented on src/JokeraceEligibility.sol:
```diff
@@ -214,27 +236,35 @@ contract JokeraceEligibility is HatsEligibilityModule {
                         INTERNAL FUNCTIONS
     //////////////////////////////////////////////////////////////*/
 
+  function _currentTermEnded(uint256 currentTermEnd) internal view returns (bool allowed) {
```
Any reason these shouldn't be public functions? Might be nice for front ends to be able to easily get these sub-components of `canStartNextTerm`

gershido replied:
Added two separate public functions: `currentTermEnded` and `nextContestCompleted`, that wrap the internal ones. The reason for the separate public functions is to make them input-less, while the internal ones receive input params to save fetches from storage (in the cases where the function that uses the internal function already fetched this param from storage for other purposes).

spengrah reviewed (approved):
approving with several suggestions. otherwise, good to go!

gershido committed:
optimizations

gershido committed:
wrap internal functions
